### PR TITLE
[Sparse] Add conversion between DGLGraph and SparseMatrix.

### LINKED ISSUE #5372: [Sparse] Convert DGLGraph to SparseMatrix.
## 🔨Work Item

**IMPORTANT:**
* This template is only for dev team to track project progress. For feature request or bug report, please use the corresponding issue templates.
* DO NOT create a new work item if the purpose is to fix an existing issue or feature request. We will directly use the issue in the project tracker.

Project tracker: https://github.com/orgs/dmlc/projects/2

### PR DESCRIPTION
**Pull request by czkkkkkk**
## Description
Resolve #5372 .

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/convert.py [hunk 1]
```diff
@@ -1043,6 +1045,56 @@ def to_homogeneous(
         return retg
 
 
+def bipartite_from_sparse(
+    sp_mat: SparseMatrix,
+    utype: str = "_U",
+    etype: str = "_E",
+    vtype: str = "_V",
+    eweight_name: str = None,
+):
```
Comment (jermainewang) [inline python/dgl/convert.py:1054]: Not sure if this is indeed needed as users can always create a DGLGraph by the rows and columns of a SparseMatrix. Note that there are also other options such as extending the `data` argument of `dgl.graph` and `dgl.heterograph` to allow SparseMatrix type. But if we do believe this is useful, let's create another PR for this new API and add proper tests/docs for it.

File: python/dgl/heterograph.py [hunk 2]
```diff
@@ -3754,44 +3755,33 @@ def out_degrees(self, u=ALL, etype=None):
...
-        """Return the adjacency matrix of edges of the given edge type.
-
-        By default, a row of returned adjacency matrix represents the
-        source of an edge and the column represents the destination.
+        return self.adj(etype)
 
-        When transpose is True, a row represents the destination and a column
-        represents the source.
+    def adj(self, etype=None, eweight_name=None):
+        """Return the sparse matrix of edges of the given edge type.
```
Comment (jermainewang) [inline python/dgl/heterograph.py:3763]:
```suggestion
        """Get the adjacency matrix of the graph.
```

File: python/dgl/heterograph.py [hunk 3]
```diff
@@ -3754,44 +3755,33 @@ def out_degrees(self, u=ALL, etype=None):
...
+        eweight_name : str, optional
+            The name of edge feature used as the non-zero values. If not given,
+            the non-zero values are all 1.
 
         Returns
         -------
-        SparseTensor or scipy.sparse.spmatrix
-            Adjacency matrix.
+        SparseMatrix
+            Sparse matrix.
```
Comment (jermainewang) [inline python/dgl/heterograph.py:3784]:
```suggestion
            The adjacency matrix.
```

File: python/dgl/heterograph.py [hunk 4]
```diff
@@ -3801,81 +3791,45 @@ def adj(self, transpose=False, ctx=F.cpu(), scipy_fmt=None, etype=None):
...
-        etid = self.get_etype_id(etype)
-        if fmt == "csc":
-            # The first two elements are number of rows and columns
-            return self._graph.adjacency_matrix_tensors(etid, True, "csr")[2:]
+        # Temporal fix to introduce a dependency on torch
+        import torch
+
+        etype = self.to_canonical_etype(etype)
+        indices = torch.stack(self.all_edges(etype=etype))
+        shape = (self.number_of_nodes(etype[0]), self.number_of_nodes(etype[2]))
```
Comment (jermainewang) [inline python/dgl/heterograph.py:3823]: Use `num_nodes` instead of `number_of_nodes`.

File: python/dgl/heterograph.py [hunk 5]
```diff
@@ -3801,81 +3791,45 @@ def adj(self, transpose=False, ctx=F.cpu(), scipy_fmt=None, etype=None):
...
+                                     [0, 1]]),
+                     values=tensor([3, 2]),
+                     shape=(2, 2), nnz=2)
         """
-        etid = self.get_etype_id(etype)
-        if fmt == "csc":
-            # The first two elements are number of rows and columns
-            return self._graph.adjacency_matrix_tensors(etid, True, "csr")[2:]
+        # Temporal fix to introduce a dependency on torch
+        import torch
```
Comment (jermainewang) [inline python/dgl/heterograph.py:3819]: Can this pass UTs of other backends? If not, it's ok to disable them.

**Review by jermainewang — CHANGES_REQUESTED** [review]

**Comment by jermainewang** [comment]
Also, once this PR is merged, let's change all the examples/tutorials to use `g.adj()` instead of `torch.stack(g.edges())`.

File: python/dgl/heterograph.py (hunk 5 above)
Comment (czkkkkkk) [reply python/dgl/heterograph.py:3819]: Skipped other backends.

File: python/dgl/convert.py (hunk 1 above)
Comment (czkkkkkk) [reply python/dgl/convert.py]: OK. I removed this interface.

**Comment by czkkkkkk** [comment]
When testing this interface, I found these a lot usages of the original `.adj()` and `.adj_sparse()`. So I renamed these two interfaces. `.adj()` is changed to `.adj_external()`, which returns sparse matrix in scipy or backend dependent sparse tensors. `.adj_sparse()` is changed to `.adj_tensors()`, which returns the tensors in COO/CSR/CSC formats.

File: python/dgl/heterograph.py [hunk 6]
```diff
@@ -3754,14 +3755,87 @@ def out_degrees(self, u=ALL, etype=None):
         else:
             return deg
 
-    def adjacency_matrix(
-        self, transpose=False, ctx=F.cpu(), scipy_fmt=None, etype=None
-    ):
+    def adjacency_matrix(self, etype=None):
```
Comment (frozenbugs) [inline python/dgl/heterograph.py:3758]: Since there is an API change, so we need to make sure we highlight this in the release note.

**Review by frozenbugs — APPROVED** [review]

File: python/dgl/heterograph.py (hunk 6 above)
Comment (czkkkkkk) [reply python/dgl/heterograph.py:3758]: OK.

File: python/dgl/heterograph.py [hunk 7]
```diff
@@ -3754,14 +3755,87 @@ def out_degrees(self, u=ALL, etype=None):
...
+            The type names of the edges. The allowed type name formats are:
+
+            * ``(str, str, str)`` for source node type, edge type and
+            destination node type.
+            * or one ``str`` edge type name if the name can uniquely identify a
+              triplet format in the graph.
+
+            Can be omitted if the graph has only one type of edges.
+
+        eweight_name : str, optional
```
[the remark this answers is not in the document]
Comment (czkkkkkk) [reply python/dgl/heterograph.py:3777]: It specifies the used edata name, which is a concept from DGLGraph. So I think it is better to use `eweight_name`.

**Review by jermainewang — APPROVED** [review]

**Comment by jermainewang** [comment]
Cool! I like the idea of renaming the old APIs to keep fallback options.

**Pull request merged**